packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -1920,19 +1898,9 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
         // the app is flushing manually, in which
         // case this invocation doesn't own
         // flushing.
-        if (this.flushMode === FlushMode.TurnBased) {
-            this.trackOrderSequentiallyCalls(callback);
-            return;
-        }
-
-        const savedFlushMode = this.flushMode;
-        this.setFlushMode(FlushMode.TurnBased);
-
-        try {
-            this.trackOrderSequentiallyCalls(callback);
+        this.trackOrderSequentiallyCalls(callback);
+        if (this.flushMode === FlushMode.Immediate) {
             this.flush();
```

Comment:
I'm not sure this code is correct. Previously, we always switched to FlushMode.TurnBased, and that ensured that ContainerRuntime.submit() would pass batch = true when calling submitMaybeChunkedMessages() (this._flushMode === FlushMode.TurnBased), which would instruct DeltaManager.submit() not to call flush() after every op.

With this change, I believe we will keep flushing ops after every op. Which is not what caller expecting.
If I got it right, it would be great to see if our UT captures this, and if not - what tests are missing to cover this case.

Plus, I believe you are breaking recursive orderSequentially() calls. flush() should be done only for outer most orderSequentially() call. Previously it was handled thanks change in flushMode and thus nested calls not calling flush(). Same feedback - we need to ensure that test coverage exists and captures this bug.

I think you can make it work (without changing all the time flushMode), but code in ContainerRuntime.submit() should take into account this._orderSequentiallyCalls when deciding if batch is on or not:
```ts
            clientSequenceNumber = this.submitMaybeChunkedMessages(
                type,
                content,
                serializedContent,
                maxOpSize,
                this._flushMode === FlushMode.TurnBased || this._orderSequentiallyCalls > 0,
                opMetadataInternal);
```

I'd also probably not do flush() here, but rather move it to trackOrderSequentiallyCalls() and do two things (if possible):
- Flush when this._orderSequentiallyCalls drops to zero (and this.flushMode === FlushMode.Immediate)
- inline trackOrderSequentiallyCalls() into orderSequentially()
- I'd simplify interface and remove batch argument at all, and fully rely on explicit flush() calls. This likely can be done, but requires care due to back-compat concerns - I'd definitely do so as a set of separate PRs.

Reply:
This check also needs to be done when setting `batch: true` metadata to the op in `submit`.
IMO its better to switch to Turnbased in orderSequentially. The state machine then is simple - TurnBased means add batch metadata and don't flush. Immediate means don't add batch metadata and flush.
This will also make it future proof and we don't need to add `this._orderSequentiallyCalls` check everywhere.